nifi-commons/nifi-data-provenance-utils/src/main/java/org/apache/nifi/provenance/AESProvenanceEventEncryptor.java

```diff
@@ -0,0 +1,237 @@
+        }
+
+        if (!SUPPORTED_VERSIONS.contains(metadata.version)) {
+            throw new EncryptionException("The event was encrypted with version " + metadata.version + " which is not in the list of supported versions " + StringUtils.join(SUPPORTED_VERSIONS, ","));
+        }
+
+        // TODO: Actually use the version to determine schema, etc.
+
+        if (keyProvider == null || !keyProvider.keyExists(metadata.keyId) || CryptoUtils.isEmpty(metadata.keyId)) {
+            throw new EncryptionException("The requested key ID " + metadata.keyId + " is not available");
+        } else {
+            try {
+                logger.debug("Decrypting provenance record " + recordId + " with key ID " + metadata.keyId);
+                EncryptionMethod method = EncryptionMethod.forAlgorithm(metadata.algorithm);
+                Cipher cipher = initCipher(method, Cipher.DECRYPT_MODE, keyProvider.getKey(metadata.keyId), metadata.ivBytes);
+
+                // Strip the metadata away to get just the cipher bytes
+                byte[] cipherBytes = extractCipherBytes(encryptedRecord, metadata);
+
+                // Perform the actual decryption
+                byte[] plainBytes = cipher.doFinal(cipherBytes);
+
+                logger.debug("Decrypted provenance event record " + recordId + " with key ID " + metadata.keyId);
+                return plainBytes;
+            } catch (EncryptionException | BadPaddingException | IllegalBlockSizeException | KeyManagementException e) {
+                final String msg = "Encountered an exception decrypting provenance record " + recordId;
+                logger.error(msg, e);
+                throw new EncryptionException(msg, e);
+            }
+        }
+    }
+
+    /**
+     * Returns a valid key identifier for this encryptor (valid for encryption and decryption) or throws an exception if none are available.
+     *
+     * @return the key ID
+     * @throws KeyManagementException if no available key IDs are valid for both operations
+     */
+    @Override
+    public String getNextKeyId() throws KeyManagementException {
+        if (keyProvider != null) {
+            List<String> availableKeyIds = keyProvider.getAvailableKeyIds();
+            if (!availableKeyIds.isEmpty()) {
+                return availableKeyIds.get(0);
+            }
+        }
+        throw new KeyManagementException("No available key IDs");
+    }
+
+    private EncryptionMetadata extractEncryptionMetadata(byte[] encryptedRecord) throws EncryptionException, IOException, ClassNotFoundException {
+        if (encryptedRecord == null || encryptedRecord.length < MIN_METADATA_LENGTH) {
+            throw new EncryptionException("The encrypted record is too short to contain the metadata");
+        }
+
+        // Skip the first byte (SENTINEL) and don't need to copy all the serialized record
+        byte[] metadataBytes = Arrays.copyOfRange(encryptedRecord, 1, encryptedRecord.length);
```

Comment:
Rather than copying the byte[] here and then wrapping in a ByteArrayInputStream, would recommend we instead just wrap the original byte[] in a ByteArrayInputStream, then call in.read() to discard the first byte -- avoids a good bit of garbage creation/collection.